### PR TITLE
[native]Catch memory allocation exception in http response processing

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Record the memory pool allocation failure (exception) of http response data in
HttpResponse and process the failure in driver execution context to fail the query
instead of causing the presto server process die
Avoid the http request get retry if the error is caused by the http response data
memory allocation failure.
Add unit tests to cover these new behavior changes.
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/Announcer.cpp [hunk 1]
```diff
@@ -147,6 +147,8 @@ void Announcer::makeAnnouncement() {
           LOG(WARNING) << "Announcement failed: HTTP "
                        << message->getStatusCode() << " - "
                        << response->dumpBodyChain();
+        } else if (response->hasError()) {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/Announcer.cpp:150]: How would this be possible?

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp [hunk 2]
```diff
@@ -103,6 +103,8 @@ void PrestoExchangeSource::doRequest() {
                   "Received HTTP {} {}",
                   headers->getStatusCode(),
                   headers->getStatusMessage()));
+        } else if (response->hasError()) {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:106]: Same question

File: presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.h [hunk 3]
```diff
@@ -31,7 +31,7 @@ class PartitionAndSerializeNode : public velox::core::PlanNode {
       uint32_t numPartitions,
       velox::RowTypePtr outputType,
       velox::core::PlanNodePtr source,
-      velox::core::PartitionFunctionFactory partitionFunctionFactory)
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.h:34]: Thanks. It would be nice to extract velox update and related changes into a separate PR to avoid coupling these this functionality changes.

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:106]: The memory allocation for the data response can throw exception: HttpResponse::append which can cause presto process die?

File: presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.h (hunk 3 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.h:34]: Sure, I just need this to make a custom build with @tanjialiang for some internal experiments.

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp [hunk 4]
```diff
@@ -206,9 +208,9 @@ void PrestoExchangeSource::processDataResponse(
 
 void PrestoExchangeSource::processDataError(
     const std::string& path,
-    const std::string& error) {
-  failedAttempts_++;
-  if (failedAttempts_ < 3) {
+    const std::string& error, bool retry) {
```
Comment (tanjialiang) [inline presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:211]: A new line for "bool retry"?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 5]
```diff
@@ -979,6 +980,60 @@ core::LocalPartitionNode::Type toLocalExchangeType(
       VELOX_UNSUPPORTED("Unsupported exchange type: {}", toJsonString(type));
   }
 }
+
+class HashPartitionFunctionSpec : public core::PartitionFunctionSpec {
```
Comment (tanjialiang) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:984]: Are the partition changes related to the http memory allocation? If not can we put them in a separate PR? Thanks!

**Review by tanjialiang — APPROVED** [review]

**Comment by tanjialiang** [comment]
Just noticed commit message spelled http wrongly
